core/state/access_witness.go

```diff
@@ -30,6 +29,9 @@ import (
 // * the second bit is set if the branch has been read
 type mode byte
 
+// UseGasFn is a function that can be used to charge gas for a given amount.
+type UseGasFn func(uint64) bool
```

Comment:
Now `Touch*(....)` functions receive a function that we'll use as a callback to charge the required gas.

Many `Touch*(...)` do more than one charging. This means that this callback could be called multiple times. 

Before I tried another idea which was not passing this callback, but asking for `availableGas`. The problem with that is that it adds a lot of strain in callers to solve partial charging. The `Touch*(...)` calls can charge for more than one thing. In the case the `availableGas` was sufficient for some of them but not all, that means the API should return:
- `gasThatMustBeCharged`
- and `ok` bool to signal if the `availableGas` was sufficient for everything or not.

The caller should:
- Always charge `gasThatMustBeCharged`, respectively of what `ok` is. Since the passed `availableGas` was at least sufficient to put _some_ stuff in the witness.
- if `ok` is false, fail due to out of gas.

The above logic should be done in all the callers. 

This proposed API where we pass a callback function, avoids all this logic in all callers. The API call will already charge for whatever was allowed, and simply return false signaling that the available gas was insufficient and let the caller fail (i.e: the caller shouldn't deal with charging stuff since that was already done in the API call).

That's the "TLDR" of this approach. I know was mentioned this still sounds like can get some pushback from the geth team, so we can try again with the other approach... but I found it a bit annoying.